### PR TITLE
Package filter in the Extensions: Manage page

### PR DESCRIPTION
**Pull request by nikosdion**
### Summary of Changes

Adds a Package filter in the Extensions: Manage page.

### Testing Instructions

Go to System, Manage, Extensions

### Actual result BEFORE applying this Pull Request

Filtering the list of extensions by the extension package ID they belong to is not possible.

### Expected result AFTER applying this Pull Request

You have a Package dropdown filter. Choosing a package shows you that package extension and all extensions Joomla knows belong to that package (based on the recorded package_id in the database).

### Documentation Changes Required

You need to document the Package filter:

* **Package**. Displays only the selected "package" type extension and all extensions which belong to it. Note: You should only uninstall the "package" type extension to prevent extension and core Joomla update problems.

### Translation Changes Required

Two translation keys were added

### Reasoning for this PR

It is very important that users only ever try to uninstall the "package" extension, not the individual extensions a package contains. Failing to do so will cause all sorts of problems:

* Trying to uninstall the package extension displays warnings about missing extensions which cannot be uninstalled.
* Some packages fail to uninstall because their installation scripts expect one or more of the included extensions to be present when the package uninstallation takes place.
* Removing all included extensions but not the package itself will find updates. Installing the updates reinstalls the missing extensions which can be confusing and disorienting for the user.
* Removing some of the extensions included in a package, for example some plugins, may cause the other extensions to fail in unpredictable ways.

Right now the only indication that an extension is part of a package is the meaningless Package ID column. Ask 100 users what that is for and 99 will have no idea.

Even worse, those of us who do know what it is cannot filter by a package ID, making it extremely hard to locate all extensions which belong to a package. We would have to search for the package name, note down the package ID, clear the filter, sort the list by package ID, start going through the pagination until we find the package ID in question in the rightmost column and carefully examine which extensions it contains. This is a very Soviet user experience, the only thing missing are the people in long coats asking us to please follow them to the back room for questioning.

With a filter like that it's easy to pinpoint the extensions included in a package, complete with the package itself. This makes management far easier **on real world sites** (default installations don't really have any packages, rendering this a moot point — but almost nobody runs Joomla like that in the real world).

### CODE REVIEW

File: administrator/components/com_installer/src/Field/PackageField.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	 * @return  array  The field option objects.
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected function getOptions()
+	{
+		$options = InstallerHelper::getPackageOptions();
+
+		return array_merge($options, parent::getOptions());
+	}
+
```
Comment (heelc29) [inline administrator/components/com_installer/src/Field/PackageField.php]:
```suggestion
```

File: administrator/components/com_installer/forms/filter_manage.xml [hunk 2]
```diff
@@ -28,6 +28,16 @@
 			<option value="">COM_INSTALLER_VALUE_CLIENT_SELECT</option>
 		</field>
 
+		<field
+			name="package_id"
+			type="package"
+			label="COM_INSTALLER_MANAGE_FILTER_PACKAGE_ID_LABEL"
+			desc="COM_INSTALLER_MANAGE_FILTER_PACKAGE_ID_DESC"
```
Comment (Kostelano) [inline administrator/components/com_installer/forms/filter_manage.xml]:
```suggestion
			description="COM_INSTALLER_MANAGE_FILTER_PACKAGE_ID_DESC"
```

Comment (nikosdion) [reply administrator/components/com_installer/forms/filter_manage.xml]: Whoops! Thanks

File: administrator/components/com_installer/src/Helper/InstallerHelper.php [hunk 3]
```diff
@@ -119,6 +120,64 @@ public static function getStateOptions()
...
+	 *
+	 * @return  array
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function getPackageOptions(): array
+	{
+		$options = [];
+
+		/** @var DatabaseDriver $db The application's database driver object */
+		$db         = Factory::getContainer()->get('DatabaseDriver');
```
Comment (heelc29) [inline administrator/components/com_installer/src/Helper/InstallerHelper.php]:
```suggestion
		$db         = Factory::getContainer()->get(DatabaseDriver::class);
```

File: administrator/components/com_installer/src/Model/ManageModel.php [hunk 4]
```diff
@@ -367,6 +369,17 @@ protected function getListQuery()
...
 
+		if ($packageId !== '')
+		{
+			$packageId = (int) $packageId;
+			$query->where(
+				'((' . $db->quoteName('package_id') . ' = :package_id) OR '
+				. '(' . $db->quoteName('extension_id') . ' = :package_id_2))'
+			)
+				->bind(':package_id', $packageId, ParameterType::INTEGER)
+				->bind(':package_id_2', $packageId, ParameterType::INTEGER);
```
Comment (heelc29) [inline administrator/components/com_installer/src/Model/ManageModel.php]:
```suggestion
			$query->where(
				'((' . $db->quoteName('package_id') . ' = :packageId1) OR '
				. '(' . $db->quoteName('extension_id') . ' = :packageId2))'
			)
				->bind([':packageId1',':packageId2'], $packageId, ParameterType::INTEGER);
```

File: administrator/components/com_installer/src/Field/PackageField.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	 * Method to get the field options.
+	 *
+	 * @return  array  The field option objects.
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected function getOptions()
+	{
+		$options = InstallerHelper::getPackageOptions();
+
+		return array_merge($options, parent::getOptions());
```
Comment (heelc29) [inline administrator/components/com_installer/src/Field/PackageField.php]:
```suggestion
		return array_merge(parent::getOptions(), $options);
```
image

File: administrator/components/com_installer/src/Helper/InstallerHelper.php [hunk 6]
```diff
@@ -119,6 +120,64 @@ public static function getStateOptions()
...
+		$query      = $db->getQuery(true)
+			->select(
+				[
+					$db->quoteName('extension_id'),
+					$db->quoteName('name'),
+					$db->quoteName('element')
+				]
+			)
```
Comment (heelc29) [inline administrator/components/com_installer/src/Helper/InstallerHelper.php]:
```suggestion
			->select(
				$db->quoteName(
					[
						'extension_id',
						'name',
						'element',
					]
				)
			)
```

**Review by heelc29 — CHANGES_REQUESTED** [review]

File: administrator/components/com_installer/src/Helper/InstallerHelper.php (hunk 3 above)
Comment (nikosdion) [reply administrator/components/com_installer/src/Helper/InstallerHelper.php]: OK, I am accepting that, it makes sense. That said, you should do a PR to Joomla itself to change how it fetches the DatabaseDriver from the container. There are far more hits of get('DatabaseDriver') than get(DatabaseDriver::class). It made me think that the _intention_ was to use the hardcoded string.

File: administrator/components/com_installer/src/Field/PackageField.php (hunk 5 above)
Comment (nikosdion) [reply administrator/components/com_installer/src/Field/PackageField.php]: I am not going to accept that on the grounds of maintaining consistency with the other filter fields of com_users and Joomla in general. Joomla puts the no selection option first on the list. Regardless of our personal preference, that's how Joomla does it, that's how we have to do it in our PRs.

Comment (heelc29) [reply administrator/components/com_installer/src/Field/PackageField.php]: Sorry, the screenshot is before the suggested change 😅

Here is for example the FolderField:
https://github.com/joomla/joomla-cms/blob/444965e0b7d5f0e393691d86d9f634989ebad383/administrator/components/com_installer/src/Field/FolderField.php#L43

Comment (nikosdion) [reply administrator/components/com_installer/src/Field/PackageField.php]: Ah, that makes more sense to me now 😄 OK, I will commit this change.

File: administrator/language/en-GB/com_installer.ini [hunk 7]
```diff
@@ -280,6 +282,7 @@ COM_INSTALLER_VALUE_CORE_SELECT="- Select Extensions -"
 COM_INSTALLER_VALUE_CORE_YES="Core Extensions"
 COM_INSTALLER_VALUE_FOLDER_NONAPPLICABLE="N/A"
 COM_INSTALLER_VALUE_FOLDER_SELECT="- Select Folder -"
+COM_INSTALLER_VALUE_PACKAGE_ID_SELECT="- Package -"
```
Comment (heelc29) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_VALUE_PACKAGE_ID_SELECT="- Select Package -"
```
image

**Comment by SumCompanyInc** [comment]
I have tested this item :white_check_mark: successfully on 7473cca7b59cc2990798ee9c0b8cf08b4c6a28d8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37838">issues.joomla.org/tracker/joomla-cms/37838</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 7473cca7b59cc2990798ee9c0b8cf08b4c6a28d8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37838">issues.joomla.org/tracker/joomla-cms/37838</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37838">issues.joomla.org/tracker/joomla-cms/37838</a>.</sub>

**Comment by bembelimen** [comment]
Nice addition, but I will move to 4.2 and ping @roland-d to merge is asap.

**Comment by nikosdion** [comment]
@bembelimen No problem and thank you!

**Comment by roland-d** [comment]
Thanks everybody

**Pull request merged**